### PR TITLE
Features for API release 

### PR DESCRIPTION
**Pull request by wanderingweights**
Here we're adding in a bunch of features to help with the release to customers. 

Includes: 

* mutli-gpu support for validators 
* validators to choose which gpu they want to use for vali 
* tagged logging to track jobs during prep task 
* allow S3 region to be passed in 
* endpoint to tell the api the current load of a validator serving organics 
* prevaluation status for jobs complete awaitin eval

### CODE REVIEW

File: validator/core/constants.py [hunk 1]
```diff
@@ -49,21 +49,23 @@
...
+MAX_SYNTH_DATA_POINTS = 100
 ADDITIONAL_SYNTH_DATA_PERCENTAGE = 1.0  # same size as training set
 
 # synth stuff
-SYNTH_GEN_BATCH_SIZE = 5
+SYNTH_GEN_BATCH_SIZE = 10
 SYNTH_MODEL_TEMPERATURE = 0.4
 CONTAINER_EVAL_RESULTS_PATH = "/aplp/evaluation_results.json"
-GPU_SERVER = os.getenv("GPU_SERVER")
+gpu_ids = os.getenv("GPU_IDS", "").strip()
```
Comment (namoray) [inline validator/core/constants.py:66]: _gpu_ids (since its a private var really)

File: validator/utils/logging.py
Comment (wanderingweights) [inline validator/utils/logging.py:1]: Ah, need to remove the print statements

File: core/models/payload_models.py [hunk 2]
```diff
@@ -110,8 +108,8 @@ class NewTaskRequest(BaseModel):
 class NewTaskResponse(BaseModel):
     success: bool = Field(..., description="Whether the task was created successfully")
     task_id: UUID | None = Field(..., description="The ID of the task")
-    created_at: datetime = Field(..., description="The creation time of the task")
-    account_id: UUID = Field(..., description="The account ID who owns the task")
+    created_at: datetime | None = Field(..., description="The creation time of the task")
```
Comment (namoray) [inline core/models/payload_models.py:111]: These should never be none if a new task is created

File: validator/core/constants.py [hunk 3]
```diff
@@ -30,12 +28,12 @@
 SUBMISSION_ENDPOINT = "/get_latest_model_submission/"
 
 # to update when live
-#PROD_CONTENT_BASE_URL = "https://content.gradients.io"
+# PROD_CONTENT_BASE_URL = "https://content.gradients.io"
```
Comment (namoray) [inline validator/core/constants.py:31]: update it (also please use # TODO: for todos)

File: validator/core/constants.py [hunk 4]
```diff
@@ -30,12 +28,12 @@
...
 DEV_CONTENT_BASE_URL = "https://dev.content.gradients.io"
 PROD_CONTENT_BASE_URL = "https://dev.content.gradients.io"
 
 # 241 is testnet
-CONTENT_BASE_URL = DEV_CONTENT_BASE_URL # if NETUID == 241 else PROD_CONTENT_BASE_URL
+CONTENT_BASE_URL = DEV_CONTENT_BASE_URL  # if NETUID == 241 else PROD_CONTENT_BASE_URL
```
Comment (namoray) [inline validator/core/constants.py:36]: Please uncomment this, and in future, dont leave this without a # TODO:
```suggestion
CONTENT_BASE_URL = DEV_CONTENT_BASE_URL  # if NETUID == 241 else PROD_CONTENT_BASE_URL
```

File: validator/core/constants.py [hunk 5]
```diff
@@ -48,22 +46,24 @@
 # task stuff
 
 
-HOW_MANY_TASKS_MINIMAL_AT_THE_SAME_TIME = 4
-NUMBER_OF_MINUTES_BETWEEN_SYNTH_TASK_CHECK = 90
+HOW_MANY_TASKS_MINIMAL_AT_THE_SAME_TIME = 3
```
Comment (namoray) [inline validator/core/constants.py:49]: What does this mean? The minimum number of tasks that can be running at the same time? dont think that makes sense

File: validator/endpoints/tasks.py [hunk 6]
```diff
@@ -197,15 +200,31 @@ async def get_leaderboard(
     return leaderboard_rows
 
 
+async def get_network_status(
+    config: Config = Depends(get_config),
+) -> TrainingTaskStatus:
+    try:
+        logger.info("IN get network status")
```
Comment (namoray) [inline validator/endpoints/tasks.py:207]: this a debug log, not info. shouldnt be there for prod

File: validator/endpoints/tasks.py [hunk 7]
```diff
@@ -197,15 +200,31 @@ async def get_leaderboard(
...
     router.add_api_route(TASKS_CREATE_ENDPOINT, create_task, methods=["POST"])
     router.add_api_route(GET_TASK_DETAILS_ENDPOINT, get_task_details, methods=["GET"])
     router.add_api_route(DELETE_TASK_ENDPOINT, delete_task, methods=["DELETE"])
     router.add_api_route(GET_TASKS_RESULTS_ENDPOINT, get_miner_breakdown, methods=["GET"])
     router.add_api_route(GET_NODE_RESULTS_ENDPOINT, get_node_results, methods=["GET"])
     router.add_api_route(GET_TASKS_BY_ACCOUNT_ENDPOINT, get_task_details_by_account, methods=["GET"])
     router.add_api_route(LEADERBOARD_ENDPOINT, get_leaderboard, methods=["GET"])
-
+    router.add_api_route(GET_NETWORK_STATUS, get_network_status, methods=["GET"])
+    logger.info("Router routes at end of factory:")
```
Comment (namoray) [inline validator/endpoints/tasks.py:227]: also a debug log, lets not leave these in

File: validator/endpoints/tasks.py [hunk 8]
```diff
@@ -60,10 +62,10 @@ async def create_task(
     end_timestamp = current_time + timedelta(hours=request.hours_to_complete)
 
     # if there are any queued jobs that are organic we can't accept any more to avoid overloading the network
-    queued_tasks = await task_sql.get_tasks_with_status(TaskStatus.DELAYED, config.psql_db, include_not_ready_tasks=True)
-    if len(queued_tasks) > 0:
-        logger.info("We already have some queued organic jobs, we can't a accept any more")
-        return NewTaskResponse(success=False, task_id=None)
+    #    queued_tasks = await task_sql.get_tasks_with_status(TaskStatus.DELAYED, config.psql_db, include_not_ready_tasks=True)
```
Comment (namoray) [inline validator/endpoints/tasks.py:65]: Why is this commented out? Does it need removing?

File: core/create_config.py [hunk 9]
```diff
@@ -98,28 +88,26 @@ def generate_validator_config(dev: bool = False) -> Dict[str, Any]:
...
-        "👀 Enter an exposed port to run the validator on (default: 9001): ") or "9001"
+    validator_port = input("👀 Enter an exposed port to run the validator on (default: 9001): ") or "9001"
+
+    gpu_ids = input("🎮 Enter comma-separated GPU IDs to use (e.g., 0,1,2, default = 0): ").strip() or "0"
 
     s3_compatible_endpoint = input("🎯 Enter s3 compatible endpoint: ")
     s3_compatible_access_key = input("🎯 Enter s3 compatible access key: ")
     s3_compatible_secret_key = input("🎯 Enter s3 compatible secret key: ")
     s3_bucket_name = input("🎯 Enter your s3 bucket name: ")
+    s3_region = input("🎯 Enter s3 region (default: us-east-1): ") or "us-east-1"
```
Comment (namoray) [inline core/create_config.py:108]: What happens if people use backblaze or non aws providers? Should they put us-east-1? is it clear from the docs?

File: validator/cycle/process_tasks.py [hunk 10]
```diff
@@ -188,15 +238,16 @@ async def _find_miners_for_task(config: Config):
...
-    try:
-        task.status = TaskStatus.PREPARING_DATA
-        await tasks_sql.update_task(task, config.psql_db)
-        task = await _run_task_prep(task, config.keypair)
-        logger.info(f"THE TASK HAS BEEN PREPPED {task}")
-        await tasks_sql.update_task(task, config.psql_db)
-    except Exception:
-        task.status = TaskStatus.PREP_TASK_FAILURE
-        await tasks_sql.update_task(task, config.psql_db)
+    async with TaskContext(str(task.task_id)):
```
Comment (namoray) [inline validator/cycle/process_tasks.py:241]: Do we need a context manager here? What are we trying to do?

File: validator/cycle/process_tasks.py [hunk 11]
```diff
@@ -253,7 +316,34 @@ async def _handle_delayed_tasks(config: Config):
     await asyncio.gather(*[_move_back_to_looking_for_nodes(task, config) for task in finished_delay_tasks])
 
 
+async def _move_to_preevaluation_status(task, config):
+    task.status = TaskStatus.PREEVALUATION
+    logger.info(f"Moving this one back {task}")
```
Comment (namoray) [inline validator/cycle/process_tasks.py:321]: back to where? 

Moving {task} from {status} to {status} if you wnata. log. shouldn't this have the extra context in?

File: validator/cycle/process_tasks.py [hunk 12]
```diff
@@ -253,7 +316,34 @@ async def _handle_delayed_tasks(config: Config):
...
+    await tasks_sql.update_task(task, config.psql_db)
+
+
+async def _move_any_evaluating_tasks_to_pending_evaluation(config: Config):
+    stopped_mid_evaluation = await tasks_sql.get_tasks_with_status(TaskStatus.EVALUATING, psql_db=config.psql_db)
+    logger.info(f"WE  ARE MOVING {len(stopped_mid_evaluation)} TASKS TO PREEVALUATION")
```
Comment (namoray) [inline validator/cycle/process_tasks.py:327]: need a consistent way to log changes in status ig?

File: validator/utils/minio.py [hunk 13]
```diff
@@ -81,9 +88,8 @@ def __del__(self):
 S3_COMPATIBLE_ENDPOINT = os.getenv("S3_COMPATIBLE_ENDPOINT", "localhost:9000")
```
Comment (namoray) [inline validator/utils/minio.py:88]: Isn't all this / shouldn't all of this be in the global config we load?

**Pull request merged**